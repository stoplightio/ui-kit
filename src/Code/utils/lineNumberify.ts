import { ASTNode } from '../types';

const NEW_LINES = /\n/g;

function countNewLines(value: string) {
  const newLines = value.match(NEW_LINES);
  if (newLines === null) return 0;
  return newLines.length;
}

function createLineNumberElement(lineNumber: number): ASTNode {
  return {
    type: 'element',
    tagName: 'span',
    properties: {
      className: ['line-number'],
    },
    children: [
      {
        type: 'text',
        value: String(lineNumber),
      },
    ],
  };
}

function createNewLineElement(): ASTNode {
  return {
    type: 'element',
    tagName: 'span',
    properties: {},
    children: [
      {
        type: 'text',
        value: '\n',
      },
    ],
  };
}

export function lineNumberify(nodes: ASTNode[]) {
  if (nodes.length === 0) return nodes;

  let lineNumber = 1;
  nodes.unshift(createLineNumberElement(lineNumber));

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (node.type !== 'text') continue;
    const newLinesAmount = countNewLines(node.value!); // text always have value
    if (newLinesAmount > 0) {
      node.value = node.value!.replace(NEW_LINES, '');
      for (let newLine = 0; newLine < newLinesAmount; newLine++) {
        lineNumber++;
        nodes.splice(i, 0, createNewLineElement(), createLineNumberElement(lineNumber));
        i += 2;
      }
    }
  }

  return nodes;
}
