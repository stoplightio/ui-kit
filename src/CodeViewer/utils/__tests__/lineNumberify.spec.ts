import * as path from 'path';

import { lineNumberify } from '../lineNumberify';

describe('lineNumberify util', () => {
  it('handles one-liners', () => {
    const ast = require(path.join(__dirname, '/fixtures/one-liner.json'));
    expect(lineNumberify(JSON.parse(JSON.stringify(ast)))).toMatchSnapshot();
  });

  it('handles multiple lines', () => {
    const ast = require(path.join(__dirname, '/fixtures/one-liner.json'));
    expect(lineNumberify(JSON.parse(JSON.stringify(ast)))).toMatchSnapshot();
  });
});
