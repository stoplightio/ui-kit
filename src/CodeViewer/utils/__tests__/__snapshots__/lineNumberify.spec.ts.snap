// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lineNumberify util handles multiple lines 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "fn",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "function",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "(",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "punctuation",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": ")",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "punctuation",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": ";",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "punctuation",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
    ],
    "properties": Object {
      "className": Array [
        "line-number",
      ],
    },
    "tagName": "span",
    "type": "element",
  },
]
`;

exports[`lineNumberify util handles one-liners 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "fn",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "function",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": "(",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "punctuation",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": ")",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "punctuation",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
      Object {
        "children": Array [
          Object {
            "type": "text",
            "value": ";",
          },
        ],
        "properties": Object {
          "className": Array [
            "token",
            "punctuation",
          ],
        },
        "tagName": "span",
        "type": "element",
      },
    ],
    "properties": Object {
      "className": Array [
        "line-number",
      ],
    },
    "tagName": "span",
    "type": "element",
  },
]
`;
